sycl/doc/design/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,1078 @@
+      return "unique id for ::(unnamed)::same_name";
+    }
+  }
+}
+namespace {
+  namespace {
+    namespace __sycl_detail {
+      static constexpr decltype(same_name) &spec_id_shim_1() {
+        // address of ::(unnamed)::(unnamed)::same_name;
+        return same_name;
+      }
+    }
+  }
+
+  namespace __sycl_detail {
+    // Sometimes we need a 'shim', which points to another 'shim' in order to
+    // "extract" a variable from an anonymous namespace unambiguosly
+    static constexpr decltype(__sycl_detail::__spec_id_shim_1()) &__spec_id_shim_2() {
+      // still address of ::(unnamed)::(unnamed)::same_name;
+      return __sycl_detail::__spec_id_shim_1();
+    }
+  }
+}
+namespace sycl {
+  namespace detail {
+    template<>
+    inline const char *get_spec_constant_symbolic_ID<::__sycl_detail::__spec_id_shim_2()>() {
+      return "unique id for ::(unnamed)::(unnamed)::same_name";
+    }
+  }
+}
+
+#include <CL/sycl/detail/spec_const_integration.hpp>
+```
+
+### DPC++ runtime
+
+For each device binary compiler generates a map
+\<Symbolic ID\> =\> \<list of spec constant descriptors\> ("ID map"). DPC++
+runtime imports that map when loading device binaries.
+It also maintains another map \<Spec const symbolic ID\> =\> \<its value\> 
+("value map") per `sycl::kernel_bundle` object. The value map is updated upon
+`kernel_bundler::set_specialization_constant<ID>(val)` and
+`handler::set_specialization_constant<ID>(val)` calls from the app.
+
+In order for runtime to access the right property, it need to compute the
+symbolic ID of a specialization constant based on user-provided inputs, such
+as non-type template argument passed to `set_specialization_constant` argument.
+DPC++ Headers section describes how symbolic IDs are generated and the same
+trick is used within `set_specialization_constant` method:
+```
+template<auto& SpecName>
+void set_specialization_constant(
+  typename std::remove_reference_t<decltype(SpecName)>::type value) {
+  const char *SymbolicID = detail::get_spec_constant_symbolic_ID();
+  // remember the value of the specialization constant
+  SpecConstantValuesMap[SymbolicID] = value;
+}
+```
+
+Before invoking JIT compilation of a program, the runtime "flushes"
+specialization constants:
+
+If native specialization constants are supported by the target device, the
+runtime iterates through the value map and invokes
+
+```
+pi_result piextProgramSetSpecializationConstant(pi_program prog,
+                                                pi_uint32 spec_id,
+                                                size_t spec_size,
+                                                const void *spec_value);
+```
+
+Plugin Interface function for descriptor of each property: `spec_id` and
+`spec_size` are taken from the descriptor, `spec_value` is calculated based on
+address of the specialization constant provided by user and `offset` field of
+the descriptor.
```

Comment:
> `spec_value` is calculated based on address of the specialization constant provided by user and `offset` field of the descriptor.

There is some missing step here.  How is `offset` used?  The value is in `specConstantValuesMap[SymbolicID]`.  Why not just pass the address of the entry in that map?

Reply:
I've tried to explain that in more details in 737abfba9dd98f7bf27b439a242e4854d2d3b7c1. Does it look better?